t/apicast-policy-headers.t

```diff
@@ -0,0 +1,510 @@
+            "configuration":
+              {
+                "request":
+                  [
+                    { "op": "add", "header": "New-Header", "value": "config_value_nh" },
+                    { "op": "add", "header": "Existing-Header", "value": "config_value_eh" }
+                  ]
+              }
+          }
+        ],
+        "api_backend": "http://test:$TEST_NGINX_SERVER_PORT/",
+        "proxy_rules": [
+          { "pattern": "/", "http_method": "GET", "metric_system_name": "hits", "delta": 2 }
+        ]
+      }
+    }
+  ]
+}
+--- backend
+  location /transactions/authrep.xml {
+    content_by_lua_block {
+      local expected = "service_token=token-value&service_id=42&usage%5Bhits%5D=2&user_key=value"
+      local args = ngx.var.args
+      if args == expected then
+        ngx.exit(200)
+      else
+        ngx.log(ngx.ERR, expected, ' did not match: ', args)
+        ngx.exit(403)
+      end
+    }
+  }
+--- upstream
+  location / {
+     content_by_lua_block {
+       local existing_header_values = ngx.req.get_headers()['Existing-Header']
+
+       if ngx.req.get_headers()['New-Header'] ~= 'config_value_nh' then
+         ngx.log(ngx.ERR, "The header 'New-Header' does not have the correct value")
+       elseif existing_header_values[1] ~= 'request_value_eh1' or
+           existing_header_values[2] ~= 'request_value_eh2' or
+           existing_header_values[3] ~= 'config_value_eh' then
+         ngx.log(ngx.ERR, "The header 'Existing-Header' does not have the correct value")
+       else
+         ngx.say('yay, api backend');
+       end
+     }
+  }
+--- request
+GET /?user_key=value
+--- more_headers
+Existing-Header: request_value_eh1
+Existing-Header: request_value_eh2
+--- response_body
+yay, api backend
+--- response_headers
+New-Header:
+Existing-Header:
+--- error_code: 200
+--- no_error_log
+[error]
+
+=== TEST 3: 'push' operation in request headers
+We test 3 things:
+1) Push op with a header that does not exist, does not change anything.
+2) Push op with a header that exists, adds a new header with the same name and
+   the given value.
+3) Check that the headers are received in the upstream but not in the response
+   of the original request.
+--- configuration
+{
+  "services": [
+    {
+      "id": 42,
+      "backend_version":  1,
+      "backend_authentication_type": "service_token",
+      "backend_authentication_value": "token-value",
+      "proxy": {
+        "policy_chain": [
+          { "name": "apicast.policy.apicast" },
+          {
+            "name": "apicast.policy.headers",
+            "configuration":
+              {
+                "request":
+                  [
+                    { "op": "push", "header": "New-Header", "value": "config_value_nh" },
+                    { "op": "push", "header": "Existing-Header", "value": "config_value_eh" }
+                  ]
+              }
+          }
+        ],
+        "api_backend": "http://test:$TEST_NGINX_SERVER_PORT/",
```

Comment:
I think it would be nice to reorganise the configuration in a way the most useful options fit the screen with the upstream/request definition. So the backend block could be gone (or moved to the top), the "api_backend" could be on the top of the config and on the bottom would be the policy_chain configuration. The next would be upstream block and the request+response. That should fit one screen so you could read the test and understand all important parts without scrolling...
Just my 2ç.

Reply:
Good point :+1: